runtime/databricks/automl_runtime/forecast/pmdarima/diagnostics.py

```diff
@@ -0,0 +1,111 @@
+#
+# Copyright (C) 2021 Databricks, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+import pandas as pd
+import numpy as np
+
+
+def generate_cutoffs(df: pd.DataFrame, horizon: int, unit: str, num_folds: int):
```

Comment:
Is this specific to pmdarima? Can we move it one level up as as a forecast util and share with prophet?

ditto for the functions below.

Reply:
Yes, the two functions used for pmdarima and prophet are similar. I didn't make them a common one because I still need to modify the pmdarima one when adding the tuning of m later (which is likely not going to be in 10.3), so I'd like to keep the existing prophet one as is for now to make sure I won't break things for 10.3.
So I'd prefer keeping the functions for pmdarim and prophet separate in this PR. WDYT?

Reply:
I feel it will be a maintenance burden in the long run if we kept them separate and have the logics diverge. IIRC, this is a generic utility for forecasting in general, and thus we should avoid the temptation to customize for different framework. We should make such utility functions universally well-defined (in terms of the forecasting problem, not in terms of individual solver), well-documented and well-tested.

Reply:
I agree that we definitely want to use common `generate_cutoffs` function (and other similar utility functions), which will also make it easier for us to extend with other forecasting models later. I have a backlog ticket (https://databricks.atlassian.net/browse/ML-19542) for cleaning up code, which I planned to do after adding the tuning for m (https://databricks.atlassian.net/browse/ML-19328). Does it sound good to make the changes with that ticket?

Reply:
I personally think we should unify early and then decide what to fork, as it'll be easier than the other way around. But I agree we don't have to do it in this PR.